packages/store-sync/src/blockLogsToStorage.ts

```diff
@@ -0,0 +1,305 @@
+        value: Value<TConfig, TTable>[TValue];
+      };
+    }[keyof Value<TConfig, TTable>];
+  }[keyof TConfig["tables"]];
+
+export type DeleteRecordOperation<TConfig extends StoreConfig> = BaseStorageOperation & {
+  type: "DeleteRecord";
+} & {
+    [TTable in keyof TConfig["tables"]]: {
+      name: TTable;
+      key: Key<TConfig, TTable>;
+    };
+  }[keyof TConfig["tables"]];
+
+export type StorageOperation<TConfig extends StoreConfig> =
+  | SetFieldOperation<TConfig>
+  | SetRecordOperation<TConfig>
+  | DeleteRecordOperation<TConfig>;
+
+export type BlockLogsToStorageOptions<TConfig extends StoreConfig = StoreConfig> = {
+  registerTables: (opts: { blockNumber: BlockLogs["blockNumber"]; tables: StoredTable[] }) => Promise<void>;
+  getTables: (opts: {
+    blockNumber: BlockLogs["blockNumber"];
+    tables: Pick<StoredTable, "address" | "namespace" | "name">[];
+  }) => Promise<StoredTable[]>;
+  storeOperations: (opts: {
+    blockNumber: BlockLogs["blockNumber"];
+    operations: StorageOperation<TConfig>[];
+  }) => Promise<void>;
+};
+
+export type BlockLogsToStorageResult<TConfig extends StoreConfig = StoreConfig> = (block: BlockLogs) => Promise<{
+  blockNumber: BlockLogs["blockNumber"];
+  operations: StorageOperation<TConfig>[];
+}>;
+
+type TableNamespace = string;
+type TableName = string;
+type TableKey = `${Address}:${TableNamespace}:${TableName}`;
+
+// hacky fix for schema registration + metadata events spanning multiple blocks
+// TODO: remove this once schema registration+metadata is one event or tx
+const visitedSchemas = new Map<TableKey, { address: Address; tableId: TableId; schema: TableSchema }>();
+const visitedMetadata = new Map<
+  TableKey,
+  { address: Address; tableId: TableId; keyNames: readonly string[]; valueNames: readonly string[] }
+>();
+
+export function blockLogsToStorage<TConfig extends StoreConfig = StoreConfig>({
+  registerTables,
+  getTables,
+  storeOperations,
+}: BlockLogsToStorageOptions<TConfig>): BlockLogsToStorageResult<TConfig> {
+  return async (block) => {
+    const newTableKeys = new Set<TableKey>();
+
+    block.logs.forEach((log) => {
+      if (log.eventName !== "StoreSetRecord") return;
+      if (log.args.table !== schemaTableId.toHex()) return;
+
+      const [tableForSchema, ...otherKeys] = log.args.key;
+      if (otherKeys.length) {
+        debug("registerSchema event is expected to have only one key in key tuple, but got multiple", log);
+      }
+
+      const tableId = TableId.fromHex(tableForSchema);
+      const schema = hexToTableSchema(log.args.data);
+
+      const key: TableKey = `${log.address}:${tableId.namespace}:${tableId.name}`;
+      if (!visitedSchemas.has(key)) {
+        visitedSchemas.set(key, { address: log.address, tableId, schema });
+        newTableKeys.add(key);
+      }
+    });
+
+    // Then find all metadata events. These should follow schema registration events and be in the same block (since they're in the same tx).
```

Comment:
seems like there is a comment missing above like "First find all schema events..." or something